R/teal_slice.R

```diff
@@ -268,22 +268,63 @@ c.teal_slice <- function(...) {
   ans
 }
 
+#' @keywords internal
+#' @export
+as.list.teal_slice <- function(x) {
+  checkmate::assert_class(x, "teal_slice")
+
+  x <- if (shiny::isRunning()) {
+    shiny::reactiveValuesToList(x)
+  } else {
+    shiny::isolate(shiny::reactiveValuesToList(x))
+  }
+
+  formals <- if (inherits(x, "teal_slice_expr")) {
+      formals(filter_expr)
+    } else {
+      formals(filter_var)
+    }
+
+  formal_args <- setdiff(names(formals), '...')
+  extra_args <- setdiff(names(x), formal_args)
+
+  x[c(formal_args, extra_args)]
+
+}
+
 
-# format method for teal_slice
 #' @export
+#' @param show_all `logical(1)` should parameters set to NULL be returned
+#' @param center `logical(1)` should the output be centered and trimmed
 #' @rdname teal_slice
 #' @keywords internal
 #'
-format.teal_slice <- function(x, show_all = FALSE, ...) {
+format.teal_slice <- function(x, show_all = FALSE, center = TRUE, ...) {
   checkmate::assert_flag(show_all)
+  checkmate::assert_flag(center)
 
-  x <- if (shiny::isRunning()) {
-    rev(shiny::reactiveValuesToList(x))
-  } else {
-    rev(shiny::isolate(shiny::reactiveValuesToList(x)))
-  }
+  x_list <- as.list(x)
+  if(!show_all) x_list <- Filter(Negate(is.null), x_list)
+
+  x_json <- jsonlite::toJSON(x_list, pretty = TRUE, auto_unbox = TRUE, digits = 16)
```

Comment:
1. I think we should go with `toJSON(null = "null")`, will save us trouble on loading.
2. I don't know what difference `auto_unbox` makes. Can you explain?

Reply:
@chlebowa auto_unbox changes arrays of length 1 into strings or integers

```{R}
> jsonlite::toJSON(as.list(x), pretty = TRUE, auto_unbox = TRUE)
{
  "dataname": "dataname1",
  "varname": "varname1",
  "choices": ["a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", "v", "w", "x", "y", "z"],
  "selected": {},
  "keep_na": true,
  "keep_inf": {},
  "fixed": false,
  "locked": false,
  "multiple": true,
  "id": "dataname1 varname1",
  "extra1": "extraone"
} 
> jsonlite::toJSON(as.list(x), pretty = TRUE, auto_unbox = FALSE)
{
  "dataname": ["dataname1"],
  "varname": ["varname1"],
  "choices": ["a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", "v", "w", "x", "y", "z"],
  "selected": {},
  "keep_na": [true],
  "keep_inf": {},
  "fixed": [false],
  "locked": [false],
  "multiple": [true],
  "id": ["dataname1 varname1"],
  "extra1": ["extraone"]
} 
```

Reply:
and for toJSON(null = "null") this was a great simplification of what I did without it so thanks!
3d3a89f5934eca3b21d91a258cb56e7ff409bf58

Reply:
> @chlebowa auto_unbox changes arrays of length 1 into strings or integers

Yes, but does it make a difference?
```
# on branch filter_panel_refactor@main
library(magrittr)
library(jsonlite)
x <- filter_var("iris", "Species", choices = character(0), sel = c("virginica", "versicolor"))
toJSON(x, null = "null", pretty = TRUE, auto_unbox = TRUE) %>% fromJSON %>% str
toJSON(x, null = "null", pretty = TRUE, auto_unbox = FALSE) %>% fromJSON %>% str
```
`auto_unbox` does not matter for converting to JSON and back (I am thinking ahead to saving and reloading). Does it matter for printing?

Reply:
If we'd ever expose a plumber API that servers teal_slices objects as JSON responses, this would matter for an automated client creation applied to our SWAGGER's OpenAPI documentation.

Reply:
I think the closest example of the difference in this JSON compared to what we have in R language, 
is using x<-c(5)[1] or x<-list(5)[[1]] instead of x<-5 when assigning x to 5.

Reply:
I understand what unboxing is, what I don't understand is what difference does it make to us in this case. I am not conviced we should worry about other languages at this time.

Also, note that we may end up with NULL values in place of "scalars" as well as "arrays": `keep_na" is a flag that can be NULL and choices is a vector that can be NULL or vector of length 0. How do you decide which cases to unbox?

Reply:
Both arrays and scalars can be NULLable. We don't necessarily need to think about other languages, but if you ever expose teal app on a shiny server that is hosted on one machine, but the frontend is started on another machine then you may need to connect those two machines with plumber APIs. Then the format of what's included in JSON matters. Imagine one url address for a teal app hosted on kubernetes+nginx that tracks the traffic (number of users) and starts multiple instances with shiny servers - this way each user has a separate session and you can handle 300 users for one teal app hosted on the same url. Those are real life applications where you separate backend from frontend and you build microservices that send POST/GET calls within which you share data in JSON format. 

If we don't want to be consistent with logic of JSON format, then we can use our own format and we can just dump what we had for previous format/print method into a text file. But in this case this would work only for our case. The applications of JSON gives us broader applications for more scenarios, more languages and a bigger tech structure. And if we are using JSON, let's just keep clean format of JSON. Having single scalar strings in arrays that never have more than one element is just filthy and lazy bad practice that brings more work for potential future users that can write tools based on our JSON format. It's not only other R packages or R shiny apps that can use our format, but Python wrappers or microservices wrapped into Docker containers. 

From our perspective of store/restore it does not have a big difference. But from a perspective of other tech environments that might read our stored structure this will have an impact. To compare how bad practice this is, it's the same as not paying attention on how you name files. You may end up with file names with spaces, dots or non-acsii characters - who cares - but would you name your file `"My file.test (1/) (1):23-34ń.pdf.pdf"` only because it does not break on your current R session on your current platform? Even though it might crash if the file is sent to a different machine, read in a different language or at a different operational system?

Reply:
If you don't have energy reading the previous post just assume that fromJSON in other languages (other tools) does not give the same output for unboxed an non-unboxed stuff as it does in R.

Reply:
I realized we should not unbox `selected` and `choices` as those can have length greater than 1 138b01265fe3cdac7f0480e5d29e92494e77cd7b

```{R}
filter_1 <- filter_var(
  dataname = "dataname1",
  varname = "varname1",
  choices = letters,
  selected = "b",
  keep_na = TRUE,
  fixed = FALSE,
  extra1 = "extraone"
)
cat(format(filter_1))
```

![image](https://github.com/insightsengineering/teal.slice/assets/133694481/89c14fb5-637b-4604-9010-49b237235aa8)

Reply:
thinking if we should get back to to_json function, as conversion to JSON now takes 3 lines

Reply:
Also I pushed some changes and included an OpenAPI YAML schema of our object stored in JSON
https://github.com/insightsengineering/teal.slice/blob/9b6260d9cb503440476d1b8147ff6fb51cd159c6/inst/teal_slices.yml

With OpanAPI YAML schema you can see which fields are mandatory, which are nullable, which are arrays, which are scalars and which objects have what type. This was created in [OpenAPI 3.1.0](https://www.openapis.org/blog/2021/02/18/openapi-specification-3-1-released) version, which is compatible with the newest [JSON 2020-12 meta-schema](https://json-schema.org/specification-links.html#2020-12)

This way, we can create clients to this JSON object in many other programming languages, making teal broadly applicable (especially through microservices). 

The current code got a bit uglier, but the current output is a clean JSON object that has it's own OpanAPI 3.1.0 YAML Schema.

```{R}
filter_1 <- filter_var(
+   dataname = "dataname1",
+   varname = "varname1",
+   choices = letters,
+   selected = "b",
+   keep_na = TRUE,
+   fixed = FALSE,
+   extra1 = "extraone"
+ )
> filter_2 <- filter_var(
+   dataname = "dataname1",
+   varname = "varname2",
+   choices = 1:10,
+   keep_na = TRUE,
+   selected = 2,
+   fixed = TRUE,
+   locked = FALSE,
+   extra2 = "extratwo"
+ )
> filter_3 <- filter_expr(
+   dataname = "data",
+   id = "FA",
+   title = "Female adults",
+   expr = "SEX == 'F' & AGE >= 18"
+ )
> 
> all_filters <- filter_settings(
+   filter_1,
+   filter_2,
+   filter_3,
+   exclude_varnames = list(
+     "dataname1" = "varname2"
+   )
+ )
> 
> all_filters_no_attrs <- filter_settings(
+   filter_1,
+   filter_3
+ )
> 
> filter_1
{
  "dataname": "dataname1",
  "varname" : "varname1",
  "choices" : ["a", "b", "c", "d", "e",...
  "selected": ["b"],
  "keep_na" : true,
  "fixed"   : false,
  "locked"  : false,
  "multiple": true,
  "id"      : "dataname1 varname1",
  "extra1"  : "extraone"
}
> 
> format(filter_1)
 [1] "{"                                                     
 [2] "  \"dataname\": \"dataname1\","                        
 [3] "  \"varname\" : \"varname1\","                         
 [4] "  \"choices\" : [\"a\", \"b\", \"c\", \"d\", \"e\",..."
 [5] "  \"selected\": [\"b\"],"                              
 [6] "  \"keep_na\" : true,"                                 
 [7] "  \"fixed\"   : false,"                                
 [8] "  \"locked\"  : false,"                                
 [9] "  \"multiple\": true,"                                 
[10] "  \"id\"      : \"dataname1 varname1\","               
[11] "  \"extra1\"  : \"extraone\""                          
[12] "}"                                                     
> 
> x <- all_filters
> x2 <- all_filters_no_attrs
> format(x)
[1] "{\n  \"slices\": [\n    {\n      \"dataname\": \"dataname1\",\n      \"varname\" : \"varname1\",\n      \"choices\" : [\"a\", \"b\", \"c\", \"d\", \"e\",...\n      \"selected\": [\"b\"],\n      \"keep_na\" : true,\n      \"fixed\"   : false,\n      \"locked\"  : false,\n      \"multiple\": true,\n      \"id\"      : \"dataname1 varname1\",\n      \"extra1\"  : \"extraone\"\n    },\n    {\n      \"dataname\": \"dataname1\",\n      \"varname\" : \"varname2\",\n      \"choices\" : [1, 2, 3, 4, 5, 6, 7, 8, ...\n      \"selected\": [2],\n      \"keep_na\" : true,\n      \"fixed\"   : true,\n      \"locked\"  : false,\n      \"multiple\": true,\n      \"id\"      : \"dataname1 varname2\",\n      \"extra2\"  : \"extratwo\"\n    },\n    {\n      \"dataname\": \"data\",\n      \"id\"      : \"FA\",\n      \"title\"   : \"Female adults\",\n      \"expr\"    : \"SEX == 'F' & AGE >= 18\",\n      \"locked\"  : false\n    }\n  ],\n  \"attributes\": {\n    \"exclude_varnames\": {\n      \"dataname1\": \"varname2\"\n    }\n  }\n}"
> format(x, show_all = TRUE)
[1] "{\n  \"slices\": [\n    {\n      \"dataname\": \"dataname1\",\n      \"varname\" : \"varname1\",\n      \"choices\" : [\"a\", \"b\", \"c\", \"d\", \"e\",...\n      \"selected\": [\"b\"],\n      \"keep_na\" : true,\n      \"keep_inf\": null,\n      \"fixed\"   : false,\n      \"locked\"  : false,\n      \"multiple\": true,\n      \"id\"      : \"dataname1 varname1\",\n      \"extra1\"  : \"extraone\"\n    },\n    {\n      \"dataname\": \"dataname1\",\n      \"varname\" : \"varname2\",\n      \"choices\" : [1, 2, 3, 4, 5, 6, 7, 8, ...\n      \"selected\": [2],\n      \"keep_na\" : true,\n      \"keep_inf\": null,\n      \"fixed\"   : true,\n      \"locked\"  : false,\n      \"multiple\": true,\n      \"id\"      : \"dataname1 varname2\",\n      \"extra2\"  : \"extratwo\"\n    },\n    {\n      \"dataname\": \"data\",\n      \"id\"      : \"FA\",\n      \"title\"   : \"Female adults\",\n      \"expr\"    : \"SEX == 'F' & AGE >= 18\",\n      \"locked\"  : false\n    }\n  ],\n  \"attributes\": {\n    \"exclude_varnames\": {\n      \"dataname1\": \"varname2\"\n    }\n  }\n}"
> print(x)
{
  "slices": [
    {
      "dataname": "dataname1",
      "varname" : "varname1",
      "choices" : ["a", "b", "c", "d", "e",...
      "selected": ["b"],
      "keep_na" : true,
      "fixed"   : false,
      "locked"  : false,
      "multiple": true,
      "id"      : "dataname1 varname1",
      "extra1"  : "extraone"
    },
    {
      "dataname": "dataname1",
      "varname" : "varname2",
      "choices" : [1, 2, 3, 4, 5, 6, 7, 8, ...
      "selected": [2],
      "keep_na" : true,
      "fixed"   : true,
      "locked"  : false,
      "multiple": true,
      "id"      : "dataname1 varname2",
      "extra2"  : "extratwo"
    },
    {
      "dataname": "data",
      "id"      : "FA",
      "title"   : "Female adults",
      "expr"    : "SEX == 'F' & AGE >= 18",
      "locked"  : false
    }
  ],
  "attributes": {
    "exclude_varnames": {
      "dataname1": "varname2"
    }
  }
} 
> print(x, show_all = TRUE)
{
  "slices": [
    {
      "dataname": "dataname1",
      "varname" : "varname1",
      "choices" : ["a", "b", "c", "d", "e",...
      "selected": ["b"],
      "keep_na" : true,
      "keep_inf": null,
      "fixed"   : false,
      "locked"  : false,
      "multiple": true,
      "id"      : "dataname1 varname1",
      "extra1"  : "extraone"
    },
    {
      "dataname": "dataname1",
      "varname" : "varname2",
      "choices" : [1, 2, 3, 4, 5, 6, 7, 8, ...
      "selected": [2],
      "keep_na" : true,
      "keep_inf": null,
      "fixed"   : true,
      "locked"  : false,
      "multiple": true,
      "id"      : "dataname1 varname2",
      "extra2"  : "extratwo"
    },
    {
      "dataname": "data",
      "id"      : "FA",
      "title"   : "Female adults",
      "expr"    : "SEX == 'F' & AGE >= 18",
      "locked"  : false
    }
  ],
  "attributes": {
    "exclude_varnames": {
      "dataname1": "varname2"
    }
  }
} 
> print(x2, show_all = TRUE)
{
  "slices": [
    {
      "dataname": "dataname1",
      "varname" : "varname1",
      "choices" : ["a", "b", "c", "d", "e",...
      "selected": ["b"],
      "keep_na" : true,
      "keep_inf": null,
      "fixed"   : false,
      "locked"  : false,
      "multiple": true,
      "id"      : "dataname1 varname1",
      "extra1"  : "extraone"
    },
    {
      "dataname": "data",
      "id"      : "FA",
      "title"   : "Female adults",
      "expr"    : "SEX == 'F' & AGE >= 18",
      "locked"  : false
    }
  ]
} 
> 
> store_slices(x, file = 'teal_slice_090.json')
> store_slices(x2, file = 'teal_slice_094.json')
> y <- restore_slices('teal_slice_090.json')
```

Reply:
thanks for the comments - final implementation with JSON including arrays was added a85455764676ada29870b01ee7c059bd564e6cd8